### DISCUSSION TITLE
Some questions about JWKS, signing algorythm and issuer certificate chain

### DISCUSSION THREAD

**Discussion by ivanov17** [authelia/authelia#6654] [Q&A] [answered]
Hello community,

Unfortunately, the OpenID Client documentation is not as clear as I would like. In particular, this concerns the signing algorythm and the issuer certificate chain. So, I have a few questions.

First, are there any restrictions or recommendations regarding the issuer key length? Could I hipothetically have problems if I use 4096 instead of 2048 for example?

Second, do I need one or more certificates in the issuer certificate chain? My understanding is that the first certificate in the chain is actually a CA. Do I need at least one more certificate? 

What information must I provide to issue certificates? Just a common name? Should the common name point to the entire Authelia domain (or maybe a subdomain of the web interface)? Or could it be any name?

Third, as far as I understand, the JWKS endpoint is open to everyone. Does it make sense in this case to store the certificate chain as a secret? Am I correct in understanding that I don't need to pass the public key in  any other way if the application can read the JWKS endpoint?

Also, if I want to use RS256 as a signing algorythm, do I really need to create a certificate chain or am I misunderstanding something?

Finally, should I pass the client secret to the application in any case? Or is this really only needed when using the HS256 algorythm? I think I should do it anyway, but I'm not entirely sure.

Thank you in advance.

**Comment by james-d-elliott** [comment, marked as answer]
Mostly per the docs:

> First, are there any restrictions or recommendations regarding the issuer key length? Could I hipothetically have problems if I use 4096 instead of 2048 for example?

![image](https://github.com/authelia/authelia/assets/3903683/2cd12931-16df-494b-bbb3-65900051740c)

As it says there, just has to be a PEM encoded RSA key with no less than 2048 bits.

> Second, do I need one or more certificates in the issuer certificate chain? My understanding is that the first certificate in the chain is actually a CA. Do I need at least one more certificate?

> Also, if I want to use RS256 as a signing algorythm, do I really need to create a certificate chain or am I misunderstanding something?

![image](https://github.com/authelia/authelia/assets/3903683/51a0da25-2af1-46c0-81ae-91597ae802fb)

You don't need any as per the arrow. If you provide one the first PEM block in the key must match the public bits for the `issuer_private_key` and all of them must be valid as highlighted, in addition the highlighted text shows if you provide more than each certificate must sequentially sign the previous one.

> What information must I provide to issue certificates? Just a common name? Should the common name point to the entire Autelia domain (or maybe a subdomain of the web interface)? Or could it be any name?

There is no restriction on our end, may be a restriction for the respective clients though it's rare.

> Third, as far as I understand, the JWKS endpoint is open to everyone. Does it make sense in this case to store the certificate chain as a secret? Am I correct in understanding that I don't need to pass the public key in any other way if the application can read the JWKS endpoint?

Doesn't matter if you do specifically.

> Finally, should I pass the client secret to the application in any case? Or is this really only needed when using the HS256 algorythm? I think I should do it anyway, but I'm not entirely sure.

The client secret is required by the client to exchange the grant for a token. This is standard practice in all OAuth 2.0 and OpenID Connect 1.0 implementations.

**Reply by ivanov17** [reply]
Thank you very much James.

> You don't need any as per the arrow. If you provide one the first PEM block in the key must match the public bits for the `issuer_private_key` and all of them must be valid as highlighted, in addition the highlighted text shows if you provide more than each certificate must sequentially sign the previous one.

Sorry, I have one more question. I would like to understand in which situation I should to provide issuer certificate chain? And in what cases do I need one certificate in the chain, and in what cases do I need two? What does this depend on?

I don't know if I even need `x5c` and `x5t` if they are not related to `RS256` algorythm. Does it depend entirely on some additional client requirements? In most cases, will the default value be enough for me? Should I be worried about this?

**Reply by james-d-elliott** [reply]
It does relate to RS256. Some clients may require a certificate which is what the x5c and x5t JWK fields provide. We have linked to the spec if you're interested.

**Reply by ivanov17** [reply]
Thank you James. I think now I understand: `x5c` and `x5t` are related to RS256, but only needed by some clients and are not required in most cases to provide RS256. Am I correct?

**Reply by james-d-elliott** [reply]
Yes

**Reply by ivanov17** [reply]
Great, thanks again for your help.
